src/plugins/dashboard/public/application/embeddable/grid/dashboard_grid.tsx

```diff
@@ -177,18 +176,17 @@ class DashboardGridUi extends React.Component<DashboardGridProps, State> {
       isLayoutInvalid,
     });
 
-    this.subscription = this.props.container
-      .getInput$()
-      .subscribe((input: DashboardContainerInput) => {
-        if (this.mounted) {
-          this.setState({
-            panels: input.panels,
-            viewMode: input.viewMode,
-            useMargins: input.useMargins,
-            expandedPanelId: input.expandedPanelId,
-          });
-        }
-      });
+    this.subscription = this.props.container.getInput$().subscribe(() => {
+      const { panels, viewMode, useMargins, expandedPanelId } = this.props.container.getInput();
+      if (this.mounted) {
+        this.setState({
+          panels,
+          viewMode,
+          useMargins,
+          expandedPanelId,
+        });
+      }
+    });
```

Comment:
This is a very strange fix for a very strange issue. The state wasn't updating properly in some cases, but only in the `dashboard_grid` component. I noticed that it wasn't behaving the same way in `dashboard_viewport` which gets the input again inside the subscription instead of using the input passed in. For some reason, the input passed in is not showing the latest state.

I will do some more digging, but for now this fix should be sufficient.